summary-sheets/02-problem-solving-paradigms.md

```diff
@@ -0,0 +1,388 @@
+# Worskshop 02: Problem Solving Paradigms
+## Overview and Motivation
+**What is a paradigm?** A paradigm is : 
+* A method of approaching and solving a specific problem
+* A guiding methodology to use when approaching a particular type pf problem
+* A tool used to attack and solve a problem
+* Crucial for competitive programming compertitions with repeated problem types
+These paradigms are very important since by using the right paradigm for a problem, we can get the **lowest Big O** runtime.
+
+We will discuss four different types of problem solving paradigms: **Complete Search** (Brute Force), **Divide and Conquer**, **The Greedy Approach**,
+and **Dynmaic Programming**.
+
+## Complete Search (Brute Force)
+### Definition
+Also known as force or recursive backtracking, The Complete Search technique is a
+method for solving a problem by traversing the entire (or part of the) search space to obtain
+the required solution.
+
+During the search, we are allowed to *prune*, e.g., choose not to explore parts of the search soace if we know that 
+that these parts cannot contain the solution.
+
+**When to use:** 
+* When there is clearly no other algorithm available (e.g. the task of enumerating all permutations
+of {0, 1, 2, . . .,N − 1} will require **_O(N!)_** regardless.
+* When better algorithms exist, but are overkill as the input size is small (e.g. N < 100)
+
+In ICPC (International Collegiate Programming Contest), Complete Search should be the first solution considered 
+as it is easy to come up with and code it. A A bug-free Complete Search solution should never receive the
+Wrong Answer (WA) response in programming contests as it explores the entire search space. However, it can excede 
+(most probably does) the time limit and therefore we need a better solution.
+
+**Example**: Given 6 < k < 13 integers, enumerate all oissible subsets of size 6 of these integers in sorted order.
+
+**Answer**: 
+~~~java
+// S is array of k integers
+ for (int a = 0; a < k - 5; a++)
+  for (int b = a += 1; b < k - 4; b++)
+    for (int c = b + 1; c < k - 3; c++)
+      for (int d = c + 1; d < k - 2; d++)
+        for (int e = d + 1; e < k - 1; e++)
+          for (int f = e + 1; f < k; f ++)
+            System.out.printf("%d %d %d %d %d %d\n", S[a], S[b], S[c], S[d], S[e], S[f]);
+~~~
+### Tips for Complete Search
+#### Tip 1:  Filtering versus Generating
+**Filters:** Programs that examine lots of (if not all) candidate solutions and choose the ones that are

```

Comment:
Can stress that it is generate and examine -> disregard incorrect solutions (this idea of throwing out *after* we've generated is what the filtration is).